xdsl/dialects/fsm.py

```diff
@@ -0,0 +1,455 @@
+            arg_names = arg_names
+            attributes["arg_names"] = arg_names
+        if res_names is not None:
+            res_names = res_names
+            attributes["res_names"] = res_names
+        if not isinstance(body, Region):
+            body = Region(Block())
+
+        super().__init__(attributes=attributes)
+
+    def verify_(self):
+        if self.arg_attrs is not None and self.arg_names is None:
+            raise VerifyException("arg_attrs must be consistent with arg_names")
+        if self.res_attrs is not None and self.res_names is None:
+            raise VerifyException("res_attrs must be consistent with res_names")
+        if self.arg_attrs is not None and self.arg_names is not None:
+            if len(self.arg_attrs) != len(self.arg_names):
+                raise VerifyException(
+                    "The number of arg_attrs and arg_names should be the same"
+                )
+        if self.res_attrs is not None and self.res_names is not None:
+            if len(self.res_attrs) != len(self.res_names):
+                raise VerifyException(
+                    "The number of res_attrs and res_names should be the same"
+                )
+
+
+@irdl_op_definition
+class Output(IRDLOperation):
+    """represents the outputs of a machine under a specific state. The
+    types of `$operands` should be consistent with the output types of the state
+    machine"""
+
+    name = "fsm.output"
+
+    operand = var_operand_def(AnyAttr())
+
+    def __init__(
+        self,
+        operand: Sequence[SSAValue | Operation],
+    ):
+        super().__init__(
+            operands=[operand],
+        )
+
+    def verify_(self):
+        parent = self.parent_op()
+        res_type_machine = type(None)
+        while HasParent(Machine):
+            if isinstance(parent, Machine):
+                res_type_machine = type(parent.arg_attrs)
+            if parent is not None:
+                parent = parent.parent_op()
+            else:
+                raise VerifyException("Output must be in a machine")
+            if res_type_machine is not type(None):
+                if not isinstance(res_type_machine, type(self.operands)):
+                    raise VerifyException(
+                        "Output type must be consistent with the machine"
+                    )
+
+
+@irdl_op_definition
+class Return(IRDLOperation):
+    """marks the end of a region of `fsm.transition` and return
+    values if the parent region is a `$guard` region"""
+
+    name = "fsm.return"
+
+    operand = operand_def(signlessIntegerLike)
+
+    def __init__(
+        self,
+        operand: SSAValue | Operation,
+    ):
+        super().__init__(
+            operands=[operand],
+        )
+
+    def verify_(self) -> None:
+        if not self.operands:
+            return
+
+        parent = self.parent_op()
+        if not isinstance(parent, Transition):
+            return
+
+        if self.parent_region() != parent.guard:
```

Comment:
I'd make the conditions explicit instead of separating the tests:

```suggestion
        if self.operands and ...:
```

Reply:
Thanks a lot for all your observations! :)

Reply:
Concerning this one I actually thought more in detail about it and probably there is no need to raise an exception

Reply:
That is fine!